File: src/sorting/quickSort.py
import secrets
from sys import stdin

rand_generator = secrets.SystemRandom()


def get_pivot_idx(begin: int, end: int) -> int:
    return rand_generator.randint(begin, end)


def partition(arr: list, left: int, right: int, pivot_idx: int) -> int:
    pivot = arr[pivot_idx]
    arr[pivot_idx], arr[right] = arr[right], arr[pivot_idx]
    i = left
    j = right - 1

    while i <= j:
        while i <= j and arr[i] <= pivot:
            i += 1
        while i <= j and pivot <= arr[j]:
            j -= 1
        if i < j:
            arr[i], arr[j] = arr[j], arr[i]
    arr[i], arr[right] = arr[right], arr[i]

    return i


def quick_sort(arr: list, left: int, right: int) -> None:
    if left >= right:
        return
    pivot_idx = get_pivot_idx(left, right)

    pivot_idx = partition(arr, left, right, pivot_idx)
    quick_sort(arr, left, pivot_idx - 1)
    quick_sort(arr, pivot_idx + 1, right)


def print_array(arr: list) -> None:
    for i in arr:
        print(' %d' % i, end='')
    print()


def get_list() -> list:
    return list(map(int, stdin.readline().split(' ')))


lst = get_list()
print_array(lst)
quick_sort(lst, 0, len(lst) - 1)
print_array(lst)


'''
Input:
5 4 3 2 1

Output:
5 4 3 2 1
1 2 3 4 5
'''

File: src/sorting/quickSort.cs
using Sorting;

namespace QuSort
{
    public class QuickSort<T> : Sorting<T> where T : IComparable<T>
    {
        public QuickSort()
        {
            _arr = Array.ConvertAll<string, T>(Console.ReadLine().Split(' ',
            StringSplitOptions.RemoveEmptyEntries), ParseType);
            _count = _arr.Length;
        }

        public QuickSort(T[] input)
        {
            _arr = input;
            _count = _arr.Length;
        }

        private int GetPivot(int left, int right)
        {
            var rand = new Random();

            return rand.Next(left, right + 1);
        }

        private int Partition(int left, int right, int pivotIdx)
        {
            if (pivotIdx != right) Swap(pivotIdx, right);

            ref T pivot = ref _arr[right];
            int i = left, j = right - 1;

            while (i <= j)
            {
                while (i <= j && pivot.CompareTo(_arr[i]) >= 0)
                    i++;
                while (i <= j && pivot.CompareTo(_arr[j]) <= 0)
                    j--;
                if (i < j) Swap(i, j);
            }
            Swap(right, i);

            return i;
        }

        private void QuSort(int left, int right)
        {
            if (left >= right) return;

            int pivotIdx = GetPivot(left, right);

            pivotIdx = Partition(left, right, pivotIdx);
            QuSort(left, pivotIdx - 1);
            QuSort(pivotIdx + 1, right);
        }

        public override void Sort()
        {
            QuSort(0, _count - 1);
        }
    }

    public static class Program
    {
        public static void Main()
        {
            var a = new QuickSort<int>();
            a.PrintArray();
            a.Sort();
            a.PrintArray();
        }
    }
}

/*
Input:
5 4 3 2 1

Output:
5 4 3 2 1
1 2 3 4 5
*/